pep-0458.txt

```diff
@@ -527,243 +789,247 @@ How Should Metadata be Generated?
+Thus, to mark a target file as part of a consistent snapshot it MUST, when
+written to disk, include its hash in its filename:
+
+  HASH.FILENAME
+    where HASH is the `hex digest`__ of the hash of the file contents and
+    FILENAME is the original filename.
+
+This means that there are multiple copies of every target file, one for each
+of the cryptographic hash functions specified above.
+
+__ https://docs.python.org/3.7/library/hashlib.html#hashlib.hash.hexdigest
+
+Assuming infinite disk space, strictly incrementing version numbers, and no
+`hash collisions`__, a client may safely read from one snapshot while PyPI
+produces another snapshot.
+
 __ https://en.wikipedia.org/wiki/Collision_(computer_science)
 
+Clients, such as pip, that use the TUF protocol MUST be modified to download
+every metadata and target file, except for *timestamp* metadata. This is done
+by including, in the file request, the version of the file (for metadata),
+or the cryptographic hash of the file (for target files) in the filename.
+
 In this simple but effective manner, PyPI is able to capture a consistent
 snapshot of all projects and the associated metadata at a given time.  The next
 subsection provides implementation details of this idea.
 
 Note: This PEP does not prohibit using advanced file systems or tools to
-produce consistent snapshots. There are two important reasons for why this PEP
-proposes the simple solution.  First, the solution does not mandate that PyPI
+produce consistent snapshots. There are two important reasons for proposing a simple solution in this PEP.
+First, the solution does not mandate that PyPI
 use any particular file system or tool.  Second, the generic file-system based
-approach allows mirrors to use extant file transfer tools such as rsync to
+approach allows mirrors to use extant file transfer tools, such as rsync, to
 efficiently transfer consistent snapshots from PyPI.
 
 
 Producing Consistent Snapshots
 ------------------------------
 
-Given a project, PyPI is responsible for updating the *bins* metadata (roles
-delegated by the *bins* role and signed with an online key).  Every project
-MUST upload its release in a single transaction.  The uploaded set of files is
-called the "project transaction".  How PyPI MAY validate the files in a project
-transaction is discussed in a later section.  For now, the focus is on how PyPI
-will respond to a project transaction.
+When a new project release is uploaded to PyPI, PyPI MUST update the *bin-n*
+metadata responsible for the target files of the project release. Remember that
+target files are sorted into bins by their filename hashes. Consequentially,
+PyPI MUST update *snapshot* to account for the updated *bin-n* metadata, and
+*timestamp* to account for the updated *snapshot* metadata. These updates
+SHOULD be handled by automated processes, e.g. one or more *transaction
+processes* and one *snapshot process*.
```

Comment:
Currently on PyPI all of this is handled in-band during the upload, and each file uploaded is treated as a separate transaction. Does this PEP require that all files for a release be uploaded within the same transaction? Are we going to have to move to an out of band process for dealing with releases and have the upload API just queue a release?